src/resize_arr.F90

```diff
@@ -0,0 +1,280 @@
+
+    type(Bank), allocatable                    :: tmp(:)
+
+    if (requested_size > current_size) then
+
+      allocate(tmp(requested_size), STAT=alloc_err)
+      tmp(1:current_size) = array
+      call move_alloc(FROM=tmp, TO=array)
+
+      current_size = requested_size
+
+    end if
+
+  end subroutine extend_array_bank
+
+!===============================================================================
+! EXTEND8_ARRAY_BANK
+!===============================================================================
+
+  subroutine extend8_array_bank(array, current_size, requested_size, alloc_err)
+
+    type(Bank), allocatable, intent(inout)     :: array(:)
+    integer(8), intent(inout)                  :: current_size
+    integer(8), intent(in)                     :: requested_size
+    integer, intent(out)                       :: alloc_err
+
+    type(Bank), allocatable                    :: tmp(:)
+
+    if (requested_size > current_size) then
+
+      allocate(tmp(requested_size), STAT=alloc_err)
+      tmp(1:current_size) = array
+      call move_alloc(FROM=tmp, TO=array)
+
+      current_size = requested_size
+
+    end if
+
+  end subroutine extend8_array_bank
+
+!===============================================================================
+! EXTEND_ARRAY_PARTICLEBUFFER
+!===============================================================================
+
+  subroutine extend_array_particlebuffer(array, current_size, requested_size, alloc_err)
+
+    type(ParticleBuffer), allocatable, intent(inout) :: array(:)
+    integer, intent(inout)                           :: current_size
+    integer, intent(in)                              :: requested_size
+    integer, intent(out)                             :: alloc_err
+
+    type(ParticleBuffer), allocatable                :: tmp(:)
+
+    if (requested_size > current_size) then
+
+      allocate(tmp(requested_size), STAT=alloc_err)
+      tmp(1:current_size) = array
+      call move_alloc(FROM=tmp, TO=array)
+
+      current_size = requested_size
+
+    end if
+
+  end subroutine extend_array_particlebuffer
+
+!===============================================================================
+! EXTEND8_ARRAY_PARTICLEBUFFER
+!===============================================================================
+
+  subroutine extend8_array_particlebuffer(array, current_size, requested_size, alloc_err)
+
+    type(ParticleBuffer), allocatable, intent(inout) :: array(:)
+    integer(8), intent(inout)                        :: current_size
+    integer(8), intent(in)                           :: requested_size
+    integer, intent(out)                             :: alloc_err
+
+    type(ParticleBuffer), allocatable                :: tmp(:)
+
+    if (requested_size > current_size) then
+
+      allocate(tmp(requested_size), STAT=alloc_err)
+      tmp(1:current_size) = array
+      call move_alloc(FROM=tmp, TO=array)
+
+      current_size = requested_size
+
+    end if
```

Comment:
I found the only difference of `extend_array ` and `resize_array ` is the former keeps the content while the latter doesn't. Maybe it is better to combine them to be one fucntion, with an additional argument `logical keep_data`, like this (yes, the `current_size` will be removed):

```
  subroutine extend_array_bank(array, new_size, keep_data, alloc_err)

    type(Bank), allocatable, intent(inout)     :: array(:)
    integer, intent(in)                        :: new_size
    logical, intent(in)                        :: keep_data
    integer, intent(out)                       :: alloc_err

    type(Bank), allocatable                    :: tmp(:)

    if (new_size < size(array)) then
      call fatal_error("Trying to EXTEND array to a smaller size.")
    end if

    if (new_size > size(array)) then
      allocate(tmp(new_size), STAT=alloc_err)
      if (keep_data) tmp(1:current_size) = array
      call move_alloc(FROM=tmp, TO=array)
    end if

  end subroutine extend_array_bank
```
What do you think? @smharper